File: backend/tools/callback_sign.js
#!/usr/bin/env node
/**
 * Generate callback headers for /agent/callback (v6.13).
 *
 * Usage:
 *  node tools/callback_sign.js --secret "$CALLBACK_SIGNATURE_SECRET" --body-json payload.json [--key-id k1]
 */
const fs = require("fs");
const crypto = require("crypto");

function parseArgs() {
  const out = {};
  const argv = process.argv.slice(2);
  for (let i=0;i<argv.length;i++){
    const a = argv[i];
    if (a.startsWith("--")) {
      const k = a.slice(2);
      const v = argv[i+1] && !argv[i+1].startsWith("--") ? argv[++i] : "true";
      out[k] = v;
    }
  }
  return out;
}

function tokenUrlSafe(n=16){
  return crypto.randomBytes(n).toString("base64url");
}

function sign(secret, ts, nonce, bodyBuf){
  const prefix = Buffer.from(`${ts}.${nonce}.`, "utf8");
  const msg = Buffer.concat([prefix, bodyBuf]);
  return crypto.createHmac("sha256", secret).update(msg).digest("hex");
}

const args = parseArgs();
const secret = args["secret"] || process.env.CALLBACK_SIGNATURE_SECRET || "";
const keyId = args["key-id"] || process.env.CALLBACK_KEY_ID || "";
const bodyPath = args["body-json"];
if (!secret) { console.error("missing --secret or CALLBACK_SIGNATURE_SECRET"); process.exit(2); }
if (!bodyPath) { console.error("missing --body-json payload.json"); process.exit(2); }

const ts = args["timestamp"] || String(Math.floor(Date.now()/1000));
const nonce = args["nonce"] || tokenUrlSafe(16);

const obj = JSON.parse(fs.readFileSync(bodyPath, "utf8"));
const body = Buffer.from(JSON.stringify(obj), "utf8");
const sig = sign(secret, ts, nonce, body);

console.log(`X-Timestamp: ${ts}`);
console.log(`X-Nonce: ${nonce}`);
console.log(`X-Signature: ${sig}`);
if (keyId) console.log(`X-Key-Id: ${keyId}`);

File: public/downloads/windows/engine/tools/callback_sign.py
#!/usr/bin/env python3
"""Generate callback headers for /agent/callback (v6.13)."""
from __future__ import annotations

import argparse
import json
import os
import secrets as _secrets
import time
import hmac
import hashlib

def sign(secret: str, timestamp: str, nonce: str, body: bytes) -> str:
    msg = (f"{timestamp}.{nonce}.").encode("utf-8") + body
    return hmac.new(secret.encode("utf-8"), msg, hashlib.sha256).hexdigest()

def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--secret", default=os.getenv("CALLBACK_SIGNATURE_SECRET",""))
    ap.add_argument("--key-id", default=os.getenv("CALLBACK_KEY_ID",""))
    ap.add_argument("--timestamp", default="")
    ap.add_argument("--nonce", default="")
    ap.add_argument("--body-json", required=True, help="Path to json file")
    args = ap.parse_args()

    if not args.secret:
        raise SystemExit("missing --secret or CALLBACK_SIGNATURE_SECRET")

    ts = args.timestamp or str(int(time.time()))
    nonce = args.nonce or _secrets.token_urlsafe(16)
    body_obj = json.load(open(args.body_json, "r", encoding="utf-8"))
    body = json.dumps(body_obj, ensure_ascii=False).encode("utf-8")
    sig = sign(args.secret, ts, nonce, body)

    print(f"X-Timestamp: {ts}")
    print(f"X-Nonce: {nonce}")
    print(f"X-Signature: {sig}")
    if args.key_id:
        print(f"X-Key-Id: {args.key_id}")
    return 0

if __name__ == "__main__":
    raise SystemExit(main())
